pennylane/templates/subroutines/qsvt.py

```diff
@@ -21,131 +21,204 @@
+            poly = np.array([-1, 0, 0.2, 0, 0.5])
+
+            hamiltonian = qml.dot([0.3, 0.4, 0.3], [qml.Z(2), qml.X(2) @ qml.Z(3), qml.X(2)])
+
+            dev = qml.device("default.qubit")
+
+            @qml.qnode(dev)
+            def circuit():
+                qml.qsvt(hamiltonian, poly, encoding_wires=[0, 1], block_encoding="prepselprep")
+                return qml.state()
+
+
+            matrix = qml.matrix(circuit, wire_order=[0, 1, 2, 3])()
+
+        .. code-block:: pycon
+
+            >>> print(np.round(matrix[:4, :4], 4).real)
+            [[-0.7158  0.      0.      0.    ]
+             [ 0.     -0.975   0.      0.    ]
+             [ 0.      0.     -0.7158  0.    ]
+             [ 0.     -0.      0.     -0.975 ]]
+
+
+        Alternatively, if the input ``A`` is a matrix, the valid values for ``block_encoding`` are
+        ``"embedding"`` and ``"fable"``. In this case, the ``encoding_wires`` parameter corresponds to
+        the ``wires`` attribute in the templates :class:`~pennylane.BlockEncode` and :class:`~pennylane.FABLE`, respectively.
+        Note that for QSVT to work, the input matrix must be Hermitian.
+
+        .. code-block:: python
+
+            # P(x) = -1 + 0.2 x^2 + 0.5 x^4
+            poly = np.array([-0.1, 0, 0.2, 0, 0.5])
+
+            A = np.array([[-0.1, 0, 0, 0.1], [0, 0.2, 0, 0], [0, 0, -0.2, -0.2], [0.1, 0, -0.2, -0.1]])
+
+            dev = qml.device("default.qubit")
+
+            @qml.qnode(dev)
+            def circuit():
+                qml.qsvt(A, poly, encoding_wires=[0, 1, 2, 3, 4], block_encoding="fable")
+                return qml.state()
+
+            matrix = qml.matrix(circuit, wire_order=[0, 1, 2, 3, 4])()
+
+        .. code-block:: pycon
+
+            >>> print(np.round(matrix[:4, :4], 4).real)
+            [[-0.0954  0.     -0.0056 -0.0054]
+             [ 0.     -0.0912 -0.     -0.    ]
+             [-0.0056  0.     -0.0788  0.0164]
+             [-0.0054 -0.      0.0164 -0.0842]]
+    """
+
+    angles = qml.math.poly_to_angles(poly, "QSVT")
     projectors = []
 
-    if convention == "Wx":
-        angles = _qsp_to_qsvt(angles)
-        global_phase = (len(angles) - 1) % 4
+    # If the input A is a Hamiltonian
+    if hasattr(A, "pauli_rep"):
+
+        if block_encoding not in ["prepselprep", "qubitization", None]:
+            raise ValueError("block_encoding should take the value 'prepselprep' or 'qubitization'")
+
+        encoding = (
+            qml.Qubitization(A, control=encoding_wires)
+            if block_encoding == "qubitization"
+            else qml.PrepSelPrep(A, control=encoding_wires)
+        )
+
+        projectors = [
+            qml.PCPhase(angle, dim=2 ** len(A.wires), wires=encoding_wires + A.wires)
+            for angle in angles
+        ]
+
+    else:
+
+        if block_encoding not in ["embedding", "fable", None]:
+            raise ValueError("block_encoding should take the value 'embedding' or 'fable'")
```

Comment:
In this error should we clarify that `block_encoding` has to be 'embedding or 'fable' when the input to A is a matrix? 

I see that it's mentioned in the **usage details**, but it might be useful to also include in the arguments.

I got this error when running the following code and it took me a second to figure out that 'prepselprep' is only for Hamiltonians.

```python
@qml.qnode(qml.device('default.qubit'))
def circuit():
    qml.qsvt(np.array([[0,0.1],[0.2,0.3]]), [200,0,1], encoding_wires=[0,1,2], block_encoding='prepselprep')
    return qml.state()

qml.matrix(circuit,wire_order=[0,1,2])()[0:2,0:2].real
```

Reply:
Also found it a bit confusing to use `block_encode = 'prepselprep'` with Hamiltonians.

This error could be clearer.

```python
H=qml.Hamiltonian([1,1], [qml.PauliX(wires=0), qml.PauliZ(wires=0)])

@qml.qnode(qml.device('default.qubit'))
def circuit():
    qml.qsvt(H, [200,0,1], encoding_wires=[0,1,2], block_encoding='prepselprep')
    return qml.state()

qml.matrix(circuit,wire_order=[0,1,2,3])()[0:2,0:2].real
```

``` ValueError: Control wires should be different from operation wires.```